src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java

```diff
@@ -1118,6 +1121,84 @@ public void insertCode(String code, boolean blockMode)
    {
       widget_.getEditor().insert(StringUtil.normalizeNewLines(code));
    }
+   
+   public void applyCodeChanges(JsdiffChange[] changes)
+   {
+      // alias apis
+      AceEditorNative editor = widget_.getEditor();
+      EditSession session = editor.getSession();
+      Selection selection = session.getSelection();
+      AceCommandManager commandManager = editor.getCommandManager();
+      
+      // function to advance the selection
+      Consumer<Integer> advanceSelection = (Integer charsLeft) -> {
+         
+         // start from current line/row
+         Position startPos = selection.getCursor();
+         
+         // iterate through rows until we've consumed all the chars
+         int row;
+         int col = startPos.getColumn();
+         for (row = startPos.getRow(); row < session.getLength(); row++) {
+            
+            // how many chars left in the current column?
+            String line = session.getLine(row);
+            int charsLeftInLine = line.length() - col;
+            
+            // is the number of chars we still need to consume lte
+            // the number of charsLeft?
+            if (charsLeft <= charsLeftInLine) 
+            {
+               col = col + charsLeft;
+               break;
+            }
+            else
+            {
+               charsLeft -= (charsLeftInLine+1); // add 1 for newline
+               col = 0;
+            }
+         }
+         
+         // move the selection
+         selection.moveCursorTo(row, col, false);
+      };
+      
+      
+      // process changes
+      for (int i = 0; i<changes.length; i++) 
+      {
+         // get change
+         JsdiffChange change = changes[i];
+         
+         // if it's the first change then set the cursor location to the beginning of the file
+         if (i == 0)
+            selection.moveCursorTo(0, 0, false);
```

Comment:
I'd just move this to before the for-loop begins.

Reply:
Unless you didn't want to check for `changes.length == 0` I guess...

Reply:
Yeah, I need to check for changes.length and not do anything to the cursor in that case. I agree though it would be more clear outside the loop.